### PR TITLE
CAS-15 Custom message input

### LINKED ISSUE #318: viewModel.sendMessage - java.lang.NullPointerException: Attempt to invoke virtual method 'java.lang.String com.getstream.sdk.chat.rest.User.getId()' on a null object reference
Got a crash testing the custom MessageInput. Using the exact same example like from the https://github.com/GetStream/stream-chat-android/blob/master/docs/MessageList.md

Crash happens when I press the send button.

```
2019-11-20 12:00:54.313 25109-25109/de.tourlina.Tourlina.staging E/AndroidRuntime: FATAL EXCEPTION: main
    Process: de.tourlina.Tourlina.staging, PID: 25109
    java.lang.NullPointerException: Attempt to invoke virtual method 'java.lang.String com.getstream.sdk.chat.rest.User.getId()' on a null object reference
        at com.getstream.sdk.chat.rest.Message.preStorage(Message.java:310)
        at com.getstream.sdk.chat.storage.StreamStorage.insertMessageForChannel(StreamStorage.java:234)
        at com.getstream.sdk.chat.viewmodel.ChannelViewModel.sendMessage(ChannelViewModel.java:981)
        at de.tourlina.tourlina.chat.CustomMessageInputView.sendMessage(CustomMessageInputView.java:127)
        at de.tourlina.tourlina.chat.CustomMessageInputView.lambda$initBinding$0$CustomMessageInputView(CustomMessageInputView.java:83)
        at de.tourlina.tourlina.chat.-$$Lambda$CustomMessageInputView$a2vpbaqRIJDpSHY8VC7IcBh0d2c.onClick(Unknown Source:2)
        at android.view.View.performClick(View.java:7140)
        at android.view.View.performClickInternal(View.java:7117)
        at android.view.View.access$3500(View.java:801)
        at android.view.View$PerformClick.run(View.java:27351)
        at android.os.Handler.handleCallback(Handler.java:883)
        at android.os.Handler.dispatchMessage(Handler.java:100)
        at android.os.Looper.loop(Looper.java:214)
        at android.app.ActivityThread.main(ActivityThread.java:7356)
        at java.lang.reflect.Method.invoke(Native Method)
        at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:492)
        at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:930)
```

### PR DESCRIPTION
**Pull request by adrian09h**
# Submit a pull request

- close #318 
- update `CustomMessageInputView`
- add `MessageInputManager`
- update ReadMe (https://github.com/GetStream/stream-chat-android/blob/fix/custom-message-input/docs/MessageInput.md#step-2-create-custommessageinputview)

In order to test, need minor change in `activity_channel.xml`

https://github.com/GetStream/stream-chat-android/blob/fix/custom-message-input/docs/MessageInput.md#step-3-edit-activity_channelxml

### CODE REVIEW

**Comment by elevenetc** [comment]
@adrian09h what's the reason of exposing ability to set edit text? Is it a feature request?
```java
    protected void setEditText(EditText editText) {
        this.editText = editText;
    }
```

**Comment by adrian09h** [comment]
> @adrian09h what's the reason of exposing ability to set edit text? Is it a feature request?
> 
> ```java
>     protected void setEditText(EditText editText) {
>         this.editText = editText;
>     }
> ```

It was my mistake I have made it for inheritance from sub class. Not need it currently., it has already `protected` :)

**Comment by elevenetc** [comment]
@adrian09h 
> It was my mistake I have made it for inheritance from sub class. Not need it currently., it has already protected :)
I mean in general, what's the purpose of exposing this `editText `? In `CustomMessageInputView` I can't see how `editText` eventually is used.

File: CHANGELOG.md [hunk 1]
```diff
@@ -28,10 +28,14 @@
 - Fix attachment duplicated issue when message edit
 - Add File Upload 2.0
 - Add editMessage function in Channel View Model
+- Add MessageInputManager
```
Comment (tbarbugli) [inline CHANGELOG.md]: @adrian09h this does not belong in this version

File: library/src/main/java/com/getstream/sdk/chat/view/MessageListView.java [hunk 2]
```diff
@@ -147,13 +149,25 @@ public void onScrolled(RecyclerView recyclerView, int dx, int dy) {
...
+                    }, 100);
                     fVPosition = currentFirstVisible;
-                    lVPosition = currentLastVisible;
                     viewModel.setThreadParentPosition(lVPosition);
                 }
             }
         });
+
+        // Lock for 500 milliseconds setMessageListScrollUp when the keyboard is displayed.
+        this.addOnLayoutChangeListener((View view, int left, int top, int right, int bottom,
```
Comment (tbarbugli) [inline library/src/main/java/com/getstream/sdk/chat/view/MessageListView.java]: @adrian09h if this is really needed, please add more information here (sounds like an hack though to do this btw)

File: docs/MessageInput.md [hunk 3]
```diff
@@ -238,154 +236,88 @@ Create your custom Message input layout named `view_custom_message_input` as sho
 </layout>
 ```
 
-#### Step 2: Create `CustomMessageInputView`
+#### Step 2: Create CustomMessageInputView
 
-As a next step we need to connect your message input to the rest of the UI.
-There a few things to think about:
+As a next step we need to extend `MessageInputView` so you can use all features by extending `MessageInputView` simply.
```
Comment (tbarbugli) [inline docs/MessageInput.md]: @adrian09h  this is not a valid english sentence :)

Comment (tbarbugli) [reply docs/MessageInput.md]: @adrian09h don't you have to mention that a MessageInputView must implement the `MessageInputManager` interface?

File: library/src/main/java/com/getstream/sdk/chat/interfaces/MessageInputManager.java [hunk 4]
```diff
@@ -0,0 +1,9 @@
+package com.getstream.sdk.chat.interfaces;
+import com.getstream.sdk.chat.rest.Message;
+
+public interface MessageInputManager {
```
Comment (tbarbugli) [inline library/src/main/java/com/getstream/sdk/chat/interfaces/MessageInputManager.java]: @adrian09h ideally a better name + a javadoc explaining what this interface is/does

**Comment by adrian09h** [comment]
@elevenetc , removed `editText` exposing as your advice.

**Comment by adrian09h** [comment]
@tbarbugli , this PR also was fixed as your comments.

File: library/src/main/java/com/getstream/sdk/chat/interfaces/MessageSendManager.java [hunk 5]
```diff
@@ -0,0 +1,12 @@
+package com.getstream.sdk.chat.interfaces;
+import com.getstream.sdk.chat.rest.Message;
+
+/**
+ * Interface for Send Message
+ */
+public interface MessageSendManager {
```
Comment (elevenetc) [inline library/src/main/java/com/getstream/sdk/chat/interfaces/MessageSendManager.java]: I would call it just listener, but not manager. An implementation is not supposed to manage anything, but just receive success/error result.

**Comment by adrian09h** [comment]
@tbarbugli , @elevenetc .  Could you please review this PR again?

File: CHANGELOG.md [hunk 6]
```diff
@@ -1,10 +1,15 @@
...
+- Add MessageSendListener
+- Update CustomMessageInputView
+- Update CustomMessageInputView2
 - Client support for anonymous and guest users
 
+
 ## December 9th, 2019 - 3.5.0
 
 - Fix set typeFace without custom font
+- Improve rendering performance for messages without Markdown markup
```
Comment (tbarbugli) [inline CHANGELOG.md]: can you remove this?

File: CHANGELOG.md [hunk 7]
```diff
@@ -1,10 +1,15 @@
 ## Upcoming
 
+- Add MessageSendListener
+- Update CustomMessageInputView
+- Update CustomMessageInputView2
```
Comment (tbarbugli) [inline CHANGELOG.md]: Not the best kind of way to explain what was added; you should explain what was added/improved (briefly) to the library and not include changes to the sample app (nobody cares about that)

**Pull request merged**